### PR TITLE
Added bloom database hash tagging analysis plugin

### PR DESCRIPTION
**Pull request by xmco**
## Description:

This PR add an hashlookup analysis plugins that query hash against a bloom file (could be the one provided by circl under the name [hashlookup](https://circl.lu/services/hashlookup/#how-to-quickly-check-a-set-of-files-in-a-local-directory) or one created by user).

This plugin allows identifying legitimate file from plaso events' hashes. By default SHA-1 are expected, as the bloom file created by CIRCL contains SHA-1, but any hash format is supported if someone is willing to create its own bloom files with uppercase SHA-256 or md5 hashes.

## Notes:

This project adds a dependency : (flor)[https://github.com/DCSO/flor], used to parse bloom files :
* This project is licensed under [BSD3](https://github.com/DCSO/flor/blob/master/setup.py), compatible with APACHE2.0 (https://www.apache.org/legal/resolved.html#category-a).
* This project is really simple and short (< 150 lines of pure python code, with no external dependency), and is compatible with Windows, Linux and MacOS.

The test file `test_data/plaso.bloom` is a specially designed bloom file with a single hash inserted.

## Checklist:
  - [x] Automated checks (Travis, Codecov, Codefactor )pass
  - [ ] No new [new dependencies](https://plaso.readthedocs.io/en/latest/sources/developer/Adding-a-new-dependency.html) are required or l2tdevtools has been updated
  - [x] Reviewer assigned

### CODE REVIEW

**Comment by joachimmetz** [comment]
Thx for the addition a couple of things first

* make this dependency optional
* PTAL at the style guide https://github.com/log2timeline/l2tdocs/blob/main/process/Style-guide.md and change the code accordingly
* remove the CIRCL.lu marketing, it looks like this analysis plugin is to support a bloom filter file, irregardless where it originates from

**Comment by xmco** [comment]
* Code has been modified to be compliant with Style Guide, I hope I didn't miss something ;
* The name has been changed to remove references to hashlookup. Nevertheless, I have left a link to the dataset provided by CIRCL in the documentation because, in my opinion, it is useful for users to have a link to a working dataset that could be used with the plugin. Also, the bloom filter files are linked to a specific implementation related to hash functions used to create the bit array or each value inserted in the database ;  
* Concerning the optional dependency, should I modify the AnalysisPluginManager in order to allow DisabledAnalysisModule in a similar way than the OutputManager as a way to manage disabled OutputModule?

**Comment by joachimmetz** [comment]
Thx I'll have a closer look a soon as time permits

> I have left a link to the dataset provided by CIRCL in the documentation because

As long if it is a single neutral reference, this should be fine. Note that the link will be removed if the reference goes "dead".

> Concerning the optional dependency, should I modify the AnalysisPluginManager in order to allow DisabledAnalysisModule in a similar way than the OutputManager as a way to manage disabled OutputModule?

I think in this case it will suffice to catch an import error in `__init__.py`

**Comment by joachimmetz** [comment]
@xmco I'll make some changes, not that I'm leaving some comments as FYI

File: plaso/analysis/bloom.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
+# -*- coding: utf-8 -*-
+"""Analysis plugin to look up file hashes in bloom database
+
```
Comment (joachimmetz) [inline plaso/analysis/bloom.py:3]: no need for this empty line

File: plaso/analysis/bloom.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+"""
+try:
```
Comment (joachimmetz) [inline plaso/analysis/bloom.py:5]: move to __init__.py

File: plaso/analysis/bloom.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+from plaso.analysis import hash_tagging
+from plaso.analysis import logger
+from plaso.analysis import manager
+
+
+class BloomAnalysisPlugin(hash_tagging.HashTaggingAnalysisPlugin):
+  """Analysis plugin for looking up hashes in bloom file."""
+
+  # hashlookup bloom files can handle a high load
```
Comment (joachimmetz) [inline plaso/analysis/bloom.py:18]: comment does not tells me anything about the code.

File: plaso/analysis/bloom.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+  # so look up all files.
+  DATA_TYPES = frozenset(['fs:stat', 'fs:stat:ntfs'])
+
+  NAME = 'bloom'
+
+  SUPPORTED_HASHES = frozenset(['sha1', 'md5', 'sha256'])
+
+  DEFAULT_LABEL = 'bloom_present'
+
```
Comment (joachimmetz) [inline plaso/analysis/bloom.py:27]: remove empty line

File: plaso/analysis/bloom.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+  def __init__(self):
+    """Initializes an hashlookup bloom  analysis plugin."""
+    super(BloomAnalysisPlugin, self).__init__()
+    self._label = self.DEFAULT_LABEL
+    self._bloom_database_path = None
+    self.bloom_filter_object = None
```
Comment (joachimmetz) [inline plaso/analysis/bloom.py:34]: @xmco why is bloom_filter_object a public class attribute?

File: plaso/analysis/bloom.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    Args:
+      hashes (list[str]): hash values to look up.
+
+    Returns:
+      list[HashAnalysis]: analysis results, or an empty list on error.
+
+    Raises:
+      RuntimeError: when the analyzer fail to get a bloom filter object
+    """
+
```
Comment (joachimmetz) [inline plaso/analysis/bloom.py:48]: no need for this empty line

File: plaso/analysis/bloom.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    Args:
+      cached (bool): should the BloomFilter be cached
+
+    Returns:
+      Optional(flor.BloomFilter): object containig a BloomFilter.
+    """
+    if self.bloom_filter_object:
+      return self.bloom_filter_object
+
+    logger.info('Open bloom database file {0:s}.'
```
Comment (joachimmetz) [inline plaso/analysis/bloom.py:88]: does not follow the conventions used elsewhere in this codebase

File: plaso/analysis/bloom.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+      logger.error('Missing optional dependency : flor')
+      return None
+
+    try:
+      bloom_filter = flor.BloomFilter()
+      with open(self._bloom_database_path, 'rb') as f:
+        bloom_filter.read(f)
+    except IOError as exception:
+      bloom_filter = None
+      logger.error('Unable to open bloom database file {0:s} with error: {1:s}.'
```
Comment (joachimmetz) [inline plaso/analysis/bloom.py:101]: `{1:s}` will fail, use `{1!s}`

File: plaso/analysis/bloom.py [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+                   .format(self._bloom_database_path, exception))
+
+    if cached:
+      self.bloom_filter_object = bloom_filter
+      return self.bloom_filter_object
+    return bloom_filter
+
+  @staticmethod
```
Comment (joachimmetz) [inline plaso/analysis/bloom.py:109]: @xmco why is this a static method?

File: plaso/analysis/bloom.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+    Args:
+      digest (str): hash to look up.
+      bloom_filter (flor.BloomFilter): instanced BloomFilter.
+
+    Returns:
+      bool: True if the hash was found, False if not or None on error.
+    """
+    value_to_test = digest.upper().encode()
+    return value_to_test in bloom_filter
+
```
Comment (joachimmetz) [inline plaso/analysis/bloom.py:122]: unnecessary empty line

File: plaso/analysis/bloom.py [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+      bloom_database_path (str): Path to the bloom file
+    """
+    self._bloom_database_path = bloom_database_path
+
+
+  def TestLoading(self):
+    """Test if the bloom file configured exist and is valid
+
+    """
+    bf = self._GetBloomFilterObject(cached=False)
```
Comment (joachimmetz) [inline plaso/analysis/bloom.py:147]: style nit: use of abbreviations

Comment (joachimmetz) [reply plaso/analysis/bloom.py:147]: Also why not `return bool(self._GetBloomFilterObject(cached=False))` ?

File: plaso/analysis/bloom.py [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+    if bf:
+      return True
+    return False
+
```
Comment (joachimmetz) [inline plaso/analysis/bloom.py:151]: missing empty line

File: requirements.txt [hunk 13]
```diff
@@ -11,6 +11,7 @@ dfdatetime >= 20221112
 dfvfs >= 20221224
 dfwinreg >= 20211207
 dtfabric >= 20220219
+flor
```
Comment (joachimmetz) [inline requirements.txt:14]: this will need to be added to dependencies.ini

Comment (joachimmetz) [reply requirements.txt:14]: Looks like Ubuntu has the 1.1.3 version - https://packages.ubuntu.com/jammy/python3-flor

No results for Fedora

**Comment by joachimmetz** [comment]
@xmco please rebase with current HEAD and apply 
[dependencies.patch.txt](https://github.com/log2timeline/plaso/files/10328743/dependencies.patch.txt)

File: requirements.txt (hunk 13 above)
Comment (william-billaud) [reply requirements.txt]: * done (c418b5935f940421f873f66a3c1aff7ebdf54276)
* need manual modification to reflect change made in e93da822c5f567a7a1a2f869ef3b83eed3fdeaf4 to [config/dpkg/control](https://github.com/log2timeline/plaso/commit/e93da822c5f567a7a1a2f869ef3b83eed3fdeaf4#diff-5e44eafdee7bfb04feaa317e5c1497264895439cd10d887e256bebf642655c95)

Comment (xmco) [reply requirements.txt]: done ( c418b5935f940421f873f66a3c1aff7ebdf54276 )
    need manual modification to reflect change made in e93da822c5f567a7a1a2f869ef3b83eed3fdeaf4 to config/dpkg/control

File: plaso/analysis/bloom.py (hunk 9 above)
Comment (xmco) [reply plaso/analysis/bloom.py]: fixed in [2ed3723](https://github.com/log2timeline/plaso/pull/4527/commits/2ed37230d22e7d89d2111b1e1ad64e2284d2b269), Sorry, I'm not really used to the Google Python style guide.

File: requirements.txt (hunk 13 above)
Comment (xmco) [reply requirements.txt]: * done in [c418b59](https://github.com/log2timeline/plaso/pull/4527/commits/c418b5935f940421f873f66a3c1aff7ebdf54276)
* Manual modification to reflect change on [config/dpkg/control](https://github.com/log2timeline/plaso/commit/e93da822c5f567a7a1a2f869ef3b83eed3fdeaf4#diff-5e44eafdee7bfb04feaa317e5c1497264895439cd10d887e256bebf642655c95)  in e93da822c5f567a7a1a2f869ef3b83eed3fdeaf4

File: plaso/analysis/bloom.py (hunk 5 above)
Comment (xmco) [reply plaso/analysis/bloom.py]: fixed in [296fa00](https://github.com/log2timeline/plaso/pull/4527/commits/296fa00ebe3109764f75b2c9adc2072c1d608fc7)

**Comment by xmco** [comment]
Hello @joachimmetz , thanks for your time.

I have applied the dependency patch and make change according to your comment.

**Comment by xmco** [comment]
I have made a mistake when applying dependency patches on requirement.txt (fixed on https://github.com/log2timeline/plaso/pull/4527/commits/5d36e6ce8c5b16da33bdfbdd1aec443c2d5a1b1b) )

**Comment by joachimmetz** [comment]
I'll have another look at it as soon as time permits

**Comment by william-billaud** [comment]
Fyi I'll be out of office until Monday January,16th.
I could answer via my phone if needed, but will not be able to push changes if necessary.

**Comment by joachimmetz** [comment]
> Fyi I'll be out of office until Monday January,16th.
> I could answer via my phone if needed, but will not be able to push changes if necessary.

thanks for the heads up, no hurry

**Comment by joachimmetz** [comment]
Note to self missing tests/cli/helpers/bloom_analysis.py

**Review by joachimmetz — APPROVED** [review]
LGTM

**Pull request merged**